src/app/curated-rolls/curatedRollService.ts

```diff
@@ -0,0 +1,167 @@
+import { D2Item, DimPlug, DimItem } from '../inventory/item-types';
+
+/**
+ * An inventory curated roll - for an item instance ID, is the item known to be curated?
+ * If it is curated, what perks are the "best"?
+ */
+export interface InventoryCuratedRoll {
+  /** Item's instance ID. */
+  id: string;
+  /** Is it a curated roll? */
+  isCuratedRoll: boolean;
+  /** What perks did the curator pick for the item? */
+  curatedPerks: number[];
+}
+
+/**
+ * Is this a weapon or armor plug that we'll consider?
+ * This is in place so that we can disregard intrinsics, shaders/cosmetics
+ * and other things (like masterworks) which add more variance than we need.
+ */
+function isWeaponOrArmorMod(plug: DimPlug): boolean {
+  if (
+    plug.plugItem.itemCategoryHashes.find(
+      (ich) =>
+        ich === 2237038328 || // intrinsics
+        ich === 945330047 || // weapon gameplay socket
+        ich === 3851138800 // armor gameplay socket
+    )
+  ) {
+    return false;
+  }
+
+  return plug.plugItem.itemCategoryHashes.some((ich) => ich === 610365472 || ich === 4104513227); // weapon or armor mod
+}
+
+/** Is the plug's hash included in the recommended perks from the curated roll? */
+function isCuratedPlug(plug: DimPlug, curatedRoll: CuratedRoll): boolean {
+  return curatedRoll.recommendedPerks.includes(plug.plugItem.hash);
+}
+
+/** Get all of the plugs for this item that match the curated roll. */
+function getCuratedPlugs(item: D2Item, curatedRoll: CuratedRoll): number[] {
+  if (!item.sockets) {
+    return [];
+  }
+
+  const curatedPlugs: number[] = [];
+
+  item.sockets.sockets.forEach((s) => {
+    if (s.plug) {
+      s.plugOptions.forEach((dp) => {
+        if (isWeaponOrArmorMod(dp) && isCuratedPlug(dp, curatedRoll)) {
+          curatedPlugs.push(dp.plugItem.hash);
+        }
+      });
+    }
+  });
+
+  return curatedPlugs;
+}
+
+/**
+ * Do all desired perks from the curated roll exist on this item?
+ * Disregards cosmetics and some other socket types.
+ */
+function allDesiredPerksExist(item: D2Item, curatedRoll: CuratedRoll): boolean {
+  if (!item.sockets) {
+    return false;
+  }
+
+  return item.sockets.sockets.every(
+    (s) =>
+      !s.plug ||
+      !isWeaponOrArmorMod(s.plug) ||
+      s.plugOptions.some((dp) => isCuratedPlug(dp, curatedRoll))
+  );
+}
+
+/** Get the inventory curated roll for this item (based off of the curated roll). */
+function getInventoryCuratedRoll(item: D2Item, curatedRoll: CuratedRoll): InventoryCuratedRoll {
+  if (!allDesiredPerksExist(item, curatedRoll)) {
+    return getNonCuratedRollIndicator(item);
+  }
+
+  return {
+    id: item.id,
+    isCuratedRoll: true,
+    curatedPerks: getCuratedPlugs(item, curatedRoll)
+  };
+}
+
+function getNonCuratedRollIndicator(item: DimItem): InventoryCuratedRoll {
+  return {
+    id: item.id,
+    isCuratedRoll: false,
+    curatedPerks: []
+  };
+}
+
+export class CuratedRollService {
+  curationEnabled: boolean;
+  private _curatedRolls: CuratedRoll[];
```

Comment:
It might be easier to store this as a map by item hash instead of as a flat array.

Reply:
You might also want to consider not storing this state here at all, and only rely on the Redux state. That'll be more consistent, and is how I want to get a lot of the other things (like Stores Service) working.

Reply:
I wasn't quite able to wrap my head around accessing Redux state from the Angular side of the house (it's used in `dimMovePopup`), which is why I store the source state in the service.

If there's a way to hook in to Redux, then we can pull the transformed external data source into a hash/array of `CuratedRoll`s and then compare the store of items against that to generate `InventoryCuratedRoll` state.

We can have multiple `CuratedRoll`s for a single item instance ID, so hashing would be a little more complex (it'd basically be the whole object, not that it's super complex).